websocket/jetty9/src/main/java/jenkins/websocket/Jetty9Provider.java

```diff
@@ -0,0 +1,149 @@
+import org.eclipse.jetty.websocket.servlet.WebSocketServletFactory;
+import org.kohsuke.MetaInfServices;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+@Restricted(NoExternalUse.class)
+@MetaInfServices(Provider.class)
+public class Jetty9Provider implements Provider {
+
+    private static final String ATTR_LISTENER = Jetty9Provider.class.getName() + ".listener";
+
+    // TODO does not seem possible to use HttpServletRequest.get/setAttribute for this
+    private static final Map<Listener, Session> sessions = new WeakHashMap<>();
+
+    private WebSocketServletFactory factory;
+
+    public Jetty9Provider() {
+        WebSocketServletFactory.class.hashCode();
+    }
+
+    private synchronized void init(HttpServletRequest req) throws Exception {
+        if (factory == null) {
+            factory = WebSocketServletFactory.Loader.load(req.getServletContext(), WebSocketPolicy.newServerPolicy());
+            factory.start();
+            factory.setCreator(Jetty9Provider::createWebSocket);
+        }
+    }
+
+    @Override
+    public Handler handle(HttpServletRequest req, HttpServletResponse rsp, Listener listener) throws Exception {
+        init(req);
+        req.setAttribute(ATTR_LISTENER, listener);
+        if (!factory.isUpgradeRequest(req, rsp)) {
+            rsp.sendError(HttpServletResponse.SC_BAD_REQUEST, "only WS connections accepted here");
+            return null;
+        }
+        if (!factory.acceptWebSocket(req, rsp)) {
+            rsp.sendError(HttpServletResponse.SC_BAD_REQUEST, "did not manage to upgrade");
+            return null;
+        }
+        return new Handler() {
+            @Override
+            public Future<Void> sendBinary(ByteBuffer data) throws IOException {
+                return session().getRemote().sendBytesByFuture(data);
+            }
+
+            @Override
+            public void sendBinary(ByteBuffer partialByte, boolean isLast) throws IOException {
+                session().getRemote().sendPartialBytes(partialByte, isLast);
+            }
+
+            @Override
+            public Future<Void> sendText(String text) throws IOException {
+                return session().getRemote().sendStringByFuture(text);
+            }
+
+            @Override
+            public void sendPing(ByteBuffer applicationData) throws IOException {
+                session().getRemote().sendPing(applicationData);
+            }
+
+            @Override
+            public void close() throws IOException {
+                session().close();
+            }
+
+            private Session session() {
+                Session session = sessions.get(listener);
+                if (session == null) {
+                    throw new IllegalStateException("missing session");
+                }
+                return session;
+            }
+        };
+    }
+
+    private static Object createWebSocket(ServletUpgradeRequest req, ServletUpgradeResponse resp) {
```

Comment:
Note the `Object` in the signature. Javadoc is super helpful

> a websocket object to use

It is more or less through trial and error that I concluded (in #4369) that a `WebSocketListener` is expected; “normal usage” apparently means adding annotations to POJOs and just remembering what method signatures to use. I think this misconceived style is inherited from the `javax.websocket` API. (Which tries to make you declare WS-aware endpoints with more annotations, and fails to provide a simple Java API call to upgrade a connection, forcing us to rely on Jetty-specific APIs.)